dds/DCPS/transport/rtps_udp/TransactionalRtpsSendQueue.cpp

```diff
@@ -0,0 +1,104 @@
+/*
+ * Distributed under the OpenDDS License.
+ * See: http://www.opendds.org/license.html
+ */
+
+#include "TransactionalRtpsSendQueue.h"
+
+OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
+
+namespace OpenDDS {
+namespace DCPS {
+
+TransactionalRtpsSendQueue::TransactionalRtpsSendQueue(ThreadStatusManager& thread_status_manager)
+  : condition_variable_(mutex_)
+  , active_transaction_count_(0)
+  , thread_status_manager_(thread_status_manager)
+{
+}
+
+bool TransactionalRtpsSendQueue::enqueue(const MetaSubmessage& ms)
+{
+  ACE_Guard<ACE_Thread_Mutex> guard(mutex_);
+  const bool empty_before = queue_.empty();
+  queue_.push_back(ms);
+  return empty_before;
+}
+
+bool TransactionalRtpsSendQueue::enqueue(const MetaSubmessageVec& vec)
+{
+  ACE_Guard<ACE_Thread_Mutex> guard(mutex_);
+  const bool empty_before = queue_.empty();
+  for (MetaSubmessageVec::const_iterator it = vec.begin(), limit = vec.end(); it != limit; ++it) {
+    queue_.push_back(*it);
+  }
+  return empty_before && !queue_.empty();
+}
+
+void TransactionalRtpsSendQueue::begin_transaction()
+{
+  ACE_Guard<ACE_Thread_Mutex> guard(mutex_);
+  ++active_transaction_count_;
+  condition_variable_.notify_one();
+}
+
+bool TransactionalRtpsSendQueue::end_transaction()
+{
+  ACE_Guard<ACE_Thread_Mutex> guard(mutex_);
+  --active_transaction_count_;
+  condition_variable_.notify_one();
+  return !queue_.empty();
+}
+
+size_t TransactionalRtpsSendQueue::active_transaction_count() const
+{
+  ACE_Guard<ACE_Thread_Mutex> guard(mutex_);
+  return active_transaction_count_;
+}
+
+void TransactionalRtpsSendQueue::swap(MetaSubmessageVec& vec)
+{
+  vec.clear();
+
+  ACE_Guard<ACE_Thread_Mutex> guard(mutex_);
+  while (active_transaction_count_ != 0) {
+    condition_variable_.wait(thread_status_manager_);
```

Comment:
This worries me, since it means that the writing thread (in this case, the transport's single EventDispatcher thread) will block, potentially preventing it from running other scheduled events. The wait will probably be small (I think the transport's reactor thread is the only thing using transactions), but this may have unintended side effects when the transport thread is particularly busy (esp within a single transaction) or as threading design changes within the transport framework.

Reply:
That's fair.  I'll change it so the periodic timer will mark the queue ready for harvest and the last transaction to finish will actually drain the queue.

Reply:
I think this puts a large number of writes (probably the majority?) back onto the transport thread, which is also not great. Perhaps if the send queue is busy (mid-transaction), the datalink just sets a flag that the next scheduling of flush_send_queue should be immediate (have zero delay) and then a successful flush can clear that flag. That should keep the writes on the eventdispatcher thread.

Reply:
I updated it so that the event thread attempts to harvest the send queue.  If successful, it will send.  If it is not successful, then other transactions are still in progress.  In this case, the other threads will harvest the queue and a `flush_send_queue` will be scheduled.  I'm using a SporadicEvent here, but the delay is always zero.  So let me know if there is a more correct way of just scheduling an event for immediate execution.

Reply:
Not sure if you saw this TSAN issue from the last run since the actions were semi-hidden for some reason: https://github.com/objectcomputing/OpenDDS/actions/runs/3242904729/jobs/5321460350 ... looks like something is maybe up with the `fsq_vec_` indexing (maybe `fsq_vec_size_` is out-of-sync with `fsq_vec_`?).

Reply:
They were.  See https://github.com/objectcomputing/OpenDDS/pull/3794/commits/dbd54c2c04aa372c804107d66e4dddaef6df4926